pkg/agent/qrm-plugins/memory/handlers/sockmem/sockmem_linux.go

```diff
@@ -0,0 +1,200 @@
+	newUpperLimit := memTotal / pageSize / 100 * uint64(tcpMemRatio)
+	if (newUpperLimit != tcpMem[2]) && (newUpperLimit > tcpMem[1]) {
+		general.Infof("write to host tcp_mem, ratio=%d, newLimit=%d, oldLimit=%d", tcpMemRatio, newUpperLimit, tcpMem[2])
+		tcpMem[2] = newUpperLimit
+		setHostTCPMemFile(hostTCPMemFile, tcpMem)
+	}
+	return nil
+}
+
+func setCg1TCPMem(podUID, containerID string, memLimit, memTCPLimit int64) error {
+	newMemTCPLimit := memLimit / 100 * int64(sockMemConfig.cgroupTCPMemRatio)
+	newMemTCPLimit = alignToPageSize(newMemTCPLimit)
+	if newMemTCPLimit < cgroupTCPMemMin2G {
+		newMemTCPLimit = cgroupTCPMemMin2G
+	} else if newMemTCPLimit >= kernSockMemAccoutingOff {
+		newMemTCPLimit = kernSockMemAccoutingOn
+	}
+
+	cgroupPath, err := cgroupcm.GetContainerRelativeCgroupPath(podUID, containerID)
+	if err != nil {
+		return err
+	}
+	if newMemTCPLimit != memTCPLimit {
+		_ = cgroupmgr.ApplyMemoryWithRelativePath(cgroupPath, &cgroupcm.MemoryData{
+			TCPMemLimitInBytes: newMemTCPLimit,
+		})
+		general.Infof("Apply TCPMemLimitInBytes: %v, old value=%d, new value=%d", cgroupPath, memTCPLimit, newMemTCPLimit)
+	}
+	return nil
+}
+
+/*
+ * SetSockMemLimit is the unified solution for tcpmem limitation.
+ * it includes 3 parts:
+ * 1, set the global tcpmem limitation by changing net.ipv4.tcp_mem.
+ * 2, do nothing under cgroupv2.
+ * 3, set the cgroup tcpmem limitation under cgroupv1.
+ */
+func SetSockMemLimit(conf *coreconfig.Configuration,
+	_ interface{}, _ *dynamicconfig.DynamicAgentConfiguration,
+	emitter metrics.MetricEmitter, metaServer *metaserver.MetaServer) {
+	general.Infof("called")
+
+	if conf == nil {
+		general.Errorf("nil extraConf")
+		return
+	} else if emitter == nil {
+		general.Errorf("nil emitter")
+		return
+	} else if metaServer == nil {
+		general.Errorf("nil metaServer")
+		return
+	}
+
+	// SettingSockMem featuregate.
+	if !conf.EnableSettingSockMem {
+		general.Infof("SetSockMemLimit disabled")
+		return
+	}
+
+	updateGlobalTCPMemRatio(conf.SetGlobalTCPMemRatio)
+	updateCgroupTCPMemRatio(conf.SetCgroupTCPMemRatio)
+	/*
+	 * Step1, set the [limit] value for host net.ipv4.tcp_mem.
+	 *
+	 * Description of net.ipv4.tcp_mem:
+	 * It includes 3 parts: min, pressure, limit.
+	 * The format is like the following:
+	 * net.ipv4.tcp_mem = [min] [pressure] [limit]
+	 *
+	 * Each parts means:
+	 * [min]: represents the minimum number of pages allowed in the queue.
+	 * [pressure]: represents the threshold at which the system considers memory
+	 *   to be under pressure due to TCP socket usage. When the memory usage reaches
+	 *   this value, the system may start taking actions like cleaning up or reclaiming memory.
+	 * [limit]: indicates the maximum number of pages allowed in the queue.
+	 */
+	// 0 means skip this feature.
+	if conf.SetGlobalTCPMemRatio != 0 {
+		_ = setHostTCPMem(metaServer.MemoryCapacity)
+	}
+	// Step2, do nothing for cg2.
+	if common.CheckCgroup2UnifiedMode() {
```

Comment:
why this strategy should be only used in v1 instead of cgroup v2?

Reply:
Because in cg2, tcpmem is accounted together with other memory. I will share a document with you for details.

Reply:
okay, pls add a brief comment in `Step2` to clarify this

Reply:
Done.